### ISSUE TITLE
Can't initialize constants from other constants

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#29098]
```dart
class Baz {
  const Baz();
}

class Bar {
  const Bar(this.baz);
  final Baz baz;
}

class Foo {
  const Foo(this.bar);
  final Bar bar;
}

const Foo foo = const Foo(const Bar(const Baz()));

const Baz baz = foo.bar.baz; // <-- why doesn't this work? 😡

void main() { }
```

**Comment by Hixie** [comment]
For what it's worth, this causes problems in Flutter at the moment. We want as many of our widgets to be instantiated `const` as possible, because it results in dramatic performance improvements (10x in some cases, because it allows us to recognise that an entire subtree hasn't changed and we can therefore shortcircuit huge tree walks). Some of these widgets are configured from the fields of other widgets. Unfortunately, even though everything is const and known at compile time, we can't actually get the data from A to B because it goes through a field accessor and this bug therefore blocks it.


> This change would now be a breaking change

_Everything_ is a breaking change. Adding a field is a breaking change (someone might have `implements`ed your class). Adding an assert is a breaking change (it might fire). Marking something as `@deprecated` is a breaking change (it has broken our build when our dependencies do that). Changing the whitespace in a comment is a breaking change (because it increases the time to download the file to just above a timeout threshold, or because someone is using `substring` to parse your source code to generate theirs).

Changing a field to a getter is a breaking change even without allowing const access through a getter, because it changes the behaviour of the member. I don't really buy this as an argument. Field/getter symmetry is already not a thing in practice.

**Comment by lrhn** [comment]
There are breaking changes and *breaking changes*. Not everything is a breaking change, or even potentially a breaking change. Refactoring is exactly about doing semantics preserving restructuring. It's based on some changes being *safe* and non-breaking.

There are things that clients of your code can reasonably depend on, and things that are not reasonable (unless you go out of your way to promise them anyway).

You can always change the behavior of your class, and then you are responsible for the breakage. if your class is intended for subclassing, or for use as a mixin, there are things you can no longer change without breaking clients that are using your class *as intended*.

If you break someone who depends on the whitespace of your code, *they* are responsible. I'm not saying it's not annoying, but you couldn't reasonably be expected to predict that.

Changing a field to a getter or vice versa, while keeping the visible behavior, is not a breaking change. That is a quite deliberate choice in the Dart language. It means that you can evolve your class without locking yourself into a specific implementation. When you can't do that, you get something like the Java "always have a private field and public get/set methods" coding style. We are trying to avoid that.

We definitely do not want to make all final fields of const classes locked into being fields. 
Adding `const` instance fields is a possible solution, and I'm sure we'll be reconsidering it again.

**Comment by Hixie** [comment]
I'm not convinced that there's a difference between a breaking change that breaks something documented as unstable and a breaking change that breaks something implied to be stable due to some language policy. In either case, your client is broken. In either case, you risk losing them.

**Comment by lrhn** [comment]
You do have a point. That is why it is so important to know which changes are *not* breaking changes in any way. Switching between a getter and field (with the same visible semantics, obviously) is always a non-breaking change. That's a good thing.

**Comment by Hixie** [comment]
> Switching between a getter and field (with the same visible semantics, obviously) is always a non-breaking change. 

It's not, actually.

Take this code:

```dart
class A {
  final int _x = 1;
  int get x => _x;
}

class B extends A {
  @override
  int get x => super.x + 1;
}

void main() {
  print(new B().x); // prints 2
}
```

...which compiles cleanly, executes cleanly, and has no analyzer warnings.

Now remove the getter and replace it with a field. The analyzer fails with a STRONG_MODE_INVALID_FIELD_OVERRIDE error.

```dart
class A {
  final int x = 1;
}

class B extends A {
  @override
  int get x => super.x + 1; // STRONG_MODE_INVALID_FIELD_OVERRIDE
}

void main() {
  print(new B().x);
}
```

**Comment by lrhn** [comment]
Yes, that is a problem with strong mode (https://github.com/dart-lang/sdk/issues/28117). It'll be fixed.

**Comment by Hixie** [comment]
Another way this could be solved, rather than the const field issue, is by allowing compile-time-evaluatable functions to be considered const. So for example if a getter just returned a field, it could be const. That would solve many other problems too, for example it would allow const asserts to do much more elaborate tests, and would allow const Widgets to be created from lists via List.generate, etc.

**Comment by lrhn** [comment]
I'm sure we can extend `const` more, but not without cost. We could make some functions markable as `const`, basically those where the body is just returning a potentially constant expression. Again, it's not something that should happen automatically, you need to opt into being const-usable, otherwise changing the function implementation will be a breaking change.

I don't think `List.generate` will ever be considered const (it contains a loop - loops and function calls are the two things that we can't generally treat as compile-time constant because we can't determined whether they will terminate).

It's a trade-off, though. Making more expression const, just because we can, isn't a goal in itself. Keeping it simple, understandable and explainable is also a goal.
